pkg/dockerregistry/server/quotarestrictedblobstore.go

```diff
@@ -0,0 +1,155 @@
+/* This module defines a wrapper for upstream blob store that does an image quota check before commiting image
+ * layer to a registry. Master server contains admission check that will refuse the manifest if the image
+ * exceeds whatever quota set. But the check occurs too late (after the layers are written). This addition
+ * allows us to refuse the layers and thus keep the storage clean.
+ *
+ * There are few things to keep in mind:
+ *
+ *      1. Origin master calculates image sizes from the contents of the layers. Registry, on the other hand,
+ *         deals with layers themselves which contain some overhead required to store file attributes, and the
+ *         layers are compressed. Thus we compare apples with peers. The check is primarily useful when the
+ *         quota is already almost reached.
+ *
+ *      2. During a push, multiple layers are uploaded. Each uploaded layer will, for a very short period of
+ *         time, raise quota usage. However, the usage will be shortly overwritten by a resource quota refresh
+ *         back to the original value because the RQ controller takes into account just images stored in
+ *         etcd, not dangling layers inside a registry. Thus, during a check, we take into account just the
+ *         single layer, not the image as a whole.
+ *
+ *         This leads to a situation where several layers can be written until a big enough will be received
+ *         that exceeds quota limit.
+ *
+ *      3. Image stream size quota doesn't accumulate. Iow, its usage is NOT stored in a resource quota object
+ *         (its always 0 (well almost - except for the very short period time noted in 2)). Therefore its
```

Comment:
Addressed, thanks!
On 02/12/2016 04:14 PM, Andy Goldstein wrote:
> In pkg/dockerregistry/server/quotarestrictedblobstore.go
> <https://github.com/openshift/origin/pull/7128#discussion_r52752757>:
>
>> + *      1. Origin master calculates image sizes from the contents of the layers. Registry, on the other hand,
>> + *         deals with layers themselves which contain some overhead required to store file attributes, and the
>> + *         layers are compressed. Thus we compare apples with peers. The check is primarily useful when the
>> + *         quota is already almost reached.
>> + *
>> + *      2. During a push, multiple layers are uploaded. Each uploaded layer will, for a very short period of
>> + *         time, raise quota usage. However, the usage will be shortly overwritten by a resource quota refresh
>> + *         back to the original value because the RQ controller takes into account just images stored in
>> + *         etcd, not dangling layers inside a registry. Thus, during a check, we take into account just the
>> + *         single layer, not the image as a whole.
>> + *
>> + *         This leads to a situation where several layers can be written until a big enough will be received
>> + *         that exceeds quota limit.
>> + *
>> + *      3. Image stream size quota doesn't accumulate. Iow, its usage is NOT stored in a resource quota object
>> + *         (its always 0 (well almost - except for the very short period time noted in 2)). Therefore its
>
> its always -> it's always
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/openshift/origin/pull/7128/files#r52752757>.
>